File: desktop-viewer/layout-inspector/presentation/src/test/kotlin/com/androidperformancestudio/desktop/CaptureArchiveCodecTest.kt
package com.androidperformancestudio.desktop

import com.androidperformancestudio.protocol.CaptureFrameCodec
import java.nio.charset.StandardCharsets
import java.nio.file.Files
import java.nio.file.Path
import java.security.MessageDigest
import java.util.zip.ZipEntry
import java.util.zip.ZipFile
import java.util.zip.ZipOutputStream
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import org.junit.jupiter.api.Assertions.assertArrayEquals
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.io.TempDir

class CaptureArchiveCodecTest {
    @TempDir
    lateinit var tempDir: Path

    private val json = Json {
        encodeDefaults = true
        prettyPrint = true
    }

    @Test
    fun `archive round trip preserves required and optional capture files`() {
        val target = tempDir.resolve("capture.apinspect")
        val input = CaptureArchivePayload(
            snapshotJson = """{"protocolVersion":{"major":1,"minor":0}}""",
            screenshotPng = byteArrayOf(0x89.toByte(), 0x50, 0x4e, 0x47),
            rawArtifacts = CaptureRawArtifacts(
                zip = byteArrayOf(0x50, 0x4b, 0x03, 0x04),
                text = "visible hierarchy",
            ),
        )
        val metadata = CaptureArchiveMetadata(
            producerVersion = "0.1.2",
            packageName = "com.androidperformancestudio.sample",
            capturedAtEpochMillis = 1234L,
            protocolMajor = 1,
            protocolMinor = 0,
        )

        val result = CaptureArchiveCodec().write(target, metadata, input)
        val output = CaptureArchiveCodec().read(target)

        assertEquals(target, result.path)
        assertTrue(result.rawArtifactsIncluded)
        assertEquals(input.snapshotJson, output.payload.snapshotJson)
        assertArrayEquals(input.screenshotPng, output.payload.screenshotPng)
        val inputRaw = requireNotNull(input.rawArtifacts)
        val outputRaw = requireNotNull(output.payload.rawArtifacts)
        assertArrayEquals(inputRaw.zip, outputRaw.zip)
        assertEquals(inputRaw.text, outputRaw.text)
        assertEquals("com.androidperformancestudio.sample", output.metadata.packageName)
        ZipFile(target.toFile()).use { zip ->
            assertEquals(
                setOf(
                    "manifest.json",
                    "capture/layout-snapshot.json",
                    "capture/screenshot.png",
                    "raw/visible-window-views.zip",
                    "raw/visible-window-views.txt",
                ),
                zip.entries().asSequence().map { it.name }.toSet(),
            )
        }
    }


    @Test
    fun `archive round trip supports layout-only captures without screenshot entry`() {
        val target = tempDir.resolve("layout-only.apinspect")
        val input = CaptureArchivePayload(
            snapshotJson = "{}",
            screenshotPng = null,
        )

        CaptureArchiveCodec().write(target, validMetadata(), input)
        val output = CaptureArchiveCodec().read(target)

        assertEquals(input.snapshotJson, output.payload.snapshotJson)
        assertEquals(null, output.payload.screenshotPng)
        ZipFile(target.toFile()).use { zip ->
            assertEquals(
                setOf(
                    "manifest.json",
                    "capture/layout-snapshot.json",
                ),
                zip.entries().asSequence().map { it.name }.toSet(),
            )
        }
    }

    @Test
    fun `archive round trip preserves optional ai analysis report`() {
        val target = tempDir.resolve("ai-report.apinspect")
        val input = CaptureArchivePayload(
            snapshotJson = "{}",
            screenshotPng = null,
            aiAnalysisReportJson = """{"summary":"ai"}""",
        )

        CaptureArchiveCodec().write(target, validMetadata(), input)
        val output = CaptureArchiveCodec().read(target)

        assertEquals(input.aiAnalysisReportJson, output.payload.aiAnalysisReportJson)
        ZipFile(target.toFile()).use { zip ->
            assertTrue(zip.entries().asSequence().any { it.name == CaptureArchivePaths.AI_ANALYSIS_REPORT })
        }
    }


    @Test
    fun `write omits oversized optional raw attachments instead of failing export`() {
        val target = tempDir.resolve("oversized-raw-text.apinspect")
        val oversizedRawText = "x".repeat(8 * 1024 * 1024 + 1)

        val result = CaptureArchiveCodec().write(
            target = target,
            metadata = validMetadata(),
            payload = CaptureArchivePayload(
                snapshotJson = "{}",
                screenshotPng = byteArrayOf(1, 2, 3),
                rawArtifacts = CaptureRawArtifacts(
                    zip = byteArrayOf(0x50, 0x4b, 0x03, 0x04),
                    text = oversizedRawText,
                ),
            ),
        )
        val output = CaptureArchiveCodec().read(target)

        assertEquals(target, result.path)
        assertEquals(false, result.rawArtifactsIncluded)
        assertEquals(null, output.payload.rawArtifacts)
        ZipFile(target.toFile()).use { zip ->
            assertEquals(
                setOf(
                    "manifest.json",
                    "capture/layout-snapshot.json",
                    "capture/screenshot.png",
                ),
                zip.entries().asSequence().map { it.name }.toSet(),
            )
        }
    }

    @Test
    fun `write and read support snapshots larger than the live transport limit`() {
        val target = tempDir.resolve("large-snapshot.apinspect")
        val snapshot = "x".repeat(CaptureFrameCodec.MAX_SNAPSHOT_BYTES + 1)

        CaptureArchiveCodec().write(
            target = target,
            metadata = validMetadata(),
            payload = CaptureArchivePayload(
                snapshotJson = snapshot,
                screenshotPng = byteArrayOf(1, 2, 3),
            ),
        )

        assertEquals(snapshot, CaptureArchiveCodec().read(target).payload.snapshotJson)
    }

    @Test
    fun `configured snapshot multiplier raises the write and read limit`() {
        val target = tempDir.resolve("adjusted-large-snapshot.apinspect")
        val limits = CaptureArchiveLimits(snapshotSizeMultiplier = 2)
        val codec = CaptureArchiveCodec(limits = limits)
        val snapshot = "x".repeat(
            CaptureArchiveLimits.BASE_MAX_SNAPSHOT_SIZE_MIB * 1024 * 1024 + 1,
        )

        codec.write(
            target = target,
            metadata = validMetadata(),
            payload = CaptureArchivePayload(
                snapshotJson = snapshot,
                screenshotPng = byteArrayOf(1, 2, 3),
            ),
        )

        assertEquals(snapshot.length, codec.read(target).payload.snapshotJson.length)
    }

    @Test
    fun `read rejects an archive without a manifest`() {
        val archive = tempDir.resolve("missing-manifest.zip")
        writeZip(
            archive,
            listOf(CaptureArchivePaths.SNAPSHOT to "{}".toByteArray()),
        )

        assertThrows(CaptureArchiveFormatException::class.java) {
            CaptureArchiveCodec().read(archive)
        }
    }

    @Test
    fun `read rejects duplicate manifest paths`() {
        val archive = tempDir.resolve("duplicate-manifest-path.zip")
        val snapshot = "{}".toByteArray()
        val screenshot = byteArrayOf(1, 2, 3)
        val manifest = validManifest(
            entries = listOf(
                manifestEntry(CaptureArchivePaths.SNAPSHOT, snapshot, required = true),
                manifestEntry(CaptureArchivePaths.SNAPSHOT, snapshot, required = true),
                manifestEntry(CaptureArchivePaths.SCREENSHOT, screenshot, required = true),
            ),
        )
        writeZip(
            archive,
            listOf(
                CaptureArchivePaths.MANIFEST to manifestBytes(manifest),
                CaptureArchivePaths.SNAPSHOT to snapshot,
                CaptureArchivePaths.SCREENSHOT to screenshot,
            ),
        )

        assertThrows(CaptureArchiveFormatException::class.java) {
            CaptureArchiveCodec().read(archive)
        }
    }

    @Test
    fun `read rejects traversal and undeclared entries`() {
        val archive = tempDir.resolve("traversal.zip")
        val snapshot = "{}".toByteArray()
        val screenshot = byteArrayOf(1, 2, 3)
        val manifest = validManifest(
            entries = listOf(
                manifestEntry(CaptureArchivePaths.SNAPSHOT, snapshot, required = true),
                manifestEntry(CaptureArchivePaths.SCREENSHOT, screenshot, required = true),
                manifestEntry("../outside.txt", byteArrayOf(9), required = false),
            ),
        )
        writeZip(
            archive,
            listOf(
                CaptureArchivePaths.MANIFEST to manifestBytes(manifest),
                CaptureArchivePaths.SNAPSHOT to snapshot,
                CaptureArchivePaths.SCREENSHOT to screenshot,
                "../outside.txt" to byteArrayOf(9),
            ),
        )

        assertThrows(CaptureArchiveFormatException::class.java) {
            CaptureArchiveCodec().read(archive)
        }

        val undeclared = createValidArchive("undeclared.zip")
        appendEntry(undeclared, "extra.bin", byteArrayOf(4))
        assertThrows(CaptureArchiveFormatException::class.java) {
            CaptureArchiveCodec().read(undeclared)
        }
    }

    @Test
    fun `read rejects checksum mismatch and incomplete raw pair`() {
        val corrupt = tempDir.resolve("corrupt.zip")
        val snapshot = "{}".toByteArray()
        val screenshot = byteArrayOf(1, 2, 3)
        val corruptManifest = validManifest(
            entries = listOf(
                manifestEntry(CaptureArchivePaths.SNAPSHOT, snapshot, required = true),
                manifestEntry(
                    CaptureArchivePaths.SCREENSHOT,
                    byteArrayOf(9, 9, 9),
                    required = true,
                ),
            ),
        )
        writeZip(
            corrupt,
            listOf(
                CaptureArchivePaths.MANIFEST to manifestBytes(corruptManifest),
                CaptureArchivePaths.SNAPSHOT to snapshot,
                CaptureArchivePaths.SCREENSHOT to screenshot,
            ),
        )
        assertThrows(CaptureArchiveFormatException::class.java) {
            CaptureArchiveCodec().read(corrupt)
        }

        val incomplete = tempDir.resolve("incomplete-raw.zip")
        val rawZip = byteArrayOf(0x50, 0x4b, 0x03, 0x04)
        val incompleteManifest = validManifest(
            entries = listOf(
                manifestEntry(CaptureArchivePaths.SNAPSHOT, snapshot, required = true),
                manifestEntry(CaptureArchivePaths.SCREENSHOT, screenshot, required = true),
                manifestEntry(CaptureArchivePaths.RAW_ZIP, rawZip, required = false),
            ),
        )
        writeZip(
            incomplete,
            listOf(
                CaptureArchivePaths.MANIFEST to manifestBytes(incompleteManifest),
                CaptureArchivePaths.SNAPSHOT to snapshot,
                CaptureArchivePaths.SCREENSHOT to screenshot,
                CaptureArchivePaths.RAW_ZIP to rawZip,
            ),
        )
        assertThrows(CaptureArchiveFormatException::class.java) {
            CaptureArchiveCodec().read(incomplete)
        }
    }

    @Test
    fun `read rejects snapshot declared above archive limit before reading content`() {
        val archive = tempDir.resolve("oversized.zip")
        val snapshot = byteArrayOf(1)
        val screenshot = byteArrayOf(1)
        val manifest = validManifest(
            entries = listOf(
                manifestEntry(CaptureArchivePaths.SNAPSHOT, snapshot, required = true).copy(
                    size = CaptureArchiveLimits.BASE_MAX_SNAPSHOT_SIZE_MIB.toLong() *
                        1024 * 1024 + 1,
                ),
                manifestEntry(CaptureArchivePaths.SCREENSHOT, screenshot, required = true),
            ),
        )
        writeZip(
            archive,
            listOf(
                CaptureArchivePaths.MANIFEST to manifestBytes(manifest),
                CaptureArchivePaths.SNAPSHOT to snapshot,
                CaptureArchivePaths.SCREENSHOT to screenshot,
            ),
        )

        assertThrows(CaptureArchiveFormatException::class.java) {
            CaptureArchiveCodec().read(archive)
        }
    }

    @Test
    fun `failed replacement preserves an existing target and removes temp files`() {
        val target = tempDir.resolve("capture.apinspect")
        Files.writeString(target, "existing")
        val codec = CaptureArchiveCodec(
            moveIntoPlace = { _, _ -> error("move failed") },
        )

        assertThrows(IllegalStateException::class.java) {
            codec.write(target, validMetadata(), validPayload())
        }

        assertEquals("existing", Files.readString(target))
        assertEquals(
            setOf("capture.apinspect"),
            Files.list(tempDir).use { paths ->
                paths.map { it.fileName.toString() }.toList().toSet()
            },
        )
    }

    private fun createValidArchive(name: String): Path {
        val archive = tempDir.resolve(name)
        CaptureArchiveCodec().write(archive, validMetadata(), validPayload())
        return archive
    }

    private fun appendEntry(
        archive: Path,
        name: String,
        bytes: ByteArray,
    ) {
        val existing = ZipFile(archive.toFile()).use { zip ->
            zip.entries().asSequence().map { entry ->
                entry.name to zip.getInputStream(entry).use { it.readAllBytes() }
            }.toList()
        }
        writeZip(archive, existing + (name to bytes))
    }

    private fun validMetadata() = CaptureArchiveMetadata(
        producerVersion = "0.1.2",
        packageName = "com.androidperformancestudio.sample",
        capturedAtEpochMillis = 1234L,
        protocolMajor = 1,
        protocolMinor = 0,
    )

    private fun validPayload() = CaptureArchivePayload(
        snapshotJson = "{}",
        screenshotPng = byteArrayOf(1, 2, 3),
    )

    private fun validManifest(
        entries: List<CaptureArchiveManifestEntry>,
    ) = CaptureArchiveManifest(
        format = CAPTURE_ARCHIVE_FORMAT,
        archiveVersion = CAPTURE_ARCHIVE_VERSION,
        producerVersion = "0.1.2",
        packageName = "com.androidperformancestudio.sample",
        capturedAtEpochMillis = 1234L,
        protocolMajor = 1,
        protocolMinor = 0,
        entries = entries,
    )

    private fun manifestEntry(
        path: String,
        bytes: ByteArray,
        required: Boolean,
    ) = CaptureArchiveManifestEntry(
        path = path,
        size = bytes.size.toLong(),
        sha256 = sha256(bytes),
        required = required,
    )

    private fun manifestBytes(manifest: CaptureArchiveManifest): ByteArray =
        json.encodeToString(manifest).toByteArray(StandardCharsets.UTF_8)

    private fun sha256(bytes: ByteArray): String =
        MessageDigest.getInstance("SHA-256")
            .digest(bytes)
            .joinToString("") { "%02x".format(it) }

    private fun writeZip(
        target: Path,
        entries: List<Pair<String, ByteArray>>,
    ) {
        ZipOutputStream(Files.newOutputStream(target)).use { output ->
            entries.forEach { (name, bytes) ->
                output.putNextEntry(ZipEntry(name))
                output.write(bytes)
                output.closeEntry()
            }
        }
    }
}

File: electron-viewer/apps/desktop/src/main/capture-archive-codec.test.ts
import { mkdtemp, readFile, writeFile } from 'node:fs/promises';
import { tmpdir } from 'node:os';
import { join } from 'node:path';
import { createHash } from 'node:crypto';
import { describe, expect, it } from 'vitest';
import {
  CAPTURE_ARCHIVE_PATHS,
  captureArchiveLimitsFor,
  CaptureArchiveCodec,
  CaptureArchiveFormatError,
  type CaptureArchiveMetadata,
} from './capture-archive-codec.js';

const metadata: CaptureArchiveMetadata = {
  producerVersion: '0.4.6', packageName: 'com.example', capturedAtEpochMillis: 1234, protocolMajor: 1, protocolMinor: 1,
};
async function target(name: string): Promise<string> { return join(await mkdtemp(join(tmpdir(), 'aps-archive-')), name); }

function crc32(bytes: Buffer): number {
  let crc = 0xffffffff;
  for (const byte of bytes) {
    crc ^= byte;
    for (let bit = 0; bit < 8; bit += 1) crc = (crc >>> 1) ^ ((crc & 1) === 0 ? 0 : 0xedb88320);
  }
  return (crc ^ 0xffffffff) >>> 0;
}

function storedZip(entries: readonly { readonly name: string; readonly bytes: Buffer }[]): Buffer {
  const localParts: Buffer[] = [];
  const centralParts: Buffer[] = [];
  let offset = 0;
  for (const entry of entries) {
    const name = Buffer.from(entry.name);
    const crc = crc32(entry.bytes);
    const local = Buffer.alloc(30);
    local.writeUInt32LE(0x04034b50); local.writeUInt16LE(20, 4); local.writeUInt16LE(0x800, 6);
    local.writeUInt32LE(crc, 14); local.writeUInt32LE(entry.bytes.length, 18); local.writeUInt32LE(entry.bytes.length, 22); local.writeUInt16LE(name.length, 26);
    localParts.push(local, name, entry.bytes);
    const central = Buffer.alloc(46);
    central.writeUInt32LE(0x02014b50); central.writeUInt16LE(20, 4); central.writeUInt16LE(20, 6); central.writeUInt16LE(0x800, 8);
    central.writeUInt32LE(crc, 16); central.writeUInt32LE(entry.bytes.length, 20); central.writeUInt32LE(entry.bytes.length, 24); central.writeUInt16LE(name.length, 28); central.writeUInt32LE(offset, 42);
    centralParts.push(central, name);
    offset += local.length + name.length + entry.bytes.length;
  }
  const centralSize = centralParts.reduce((sum, part) => sum + part.length, 0);
  const end = Buffer.alloc(22);
  end.writeUInt32LE(0x06054b50); end.writeUInt16LE(entries.length, 8); end.writeUInt16LE(entries.length, 10); end.writeUInt32LE(centralSize, 12); end.writeUInt32LE(offset, 16);
  return Buffer.concat([...localParts, ...centralParts, end]);
}

function manifest(entries: readonly { readonly path: string; readonly bytes: Buffer; readonly required?: boolean }[], archiveVersion = 1): Buffer {
  return Buffer.from(JSON.stringify({
    format: 'agentperf-inspector-capture', archiveVersion, ...metadata,
    entries: entries.map((entry) => ({ path: entry.path, size: entry.bytes.length, sha256: createHash('sha256').update(entry.bytes).digest('hex'), required: entry.required ?? false })),
  }));
}

async function writeArchive(path: string, entries: readonly { readonly path: string; readonly bytes: Buffer; readonly required?: boolean }[], archiveVersion = 1): Promise<void> {
  await writeFile(path, storedZip([{ name: 'manifest.json', bytes: manifest(entries, archiveVersion) }, ...entries.map((entry) => ({ name: entry.path, bytes: entry.bytes }))]));
}

describe('CaptureArchiveCodec', () => {
  it('derives the Kotlin archive limits from the persisted multiplier', () => {
    const mib = 1024 * 1024;
    expect(captureArchiveLimitsFor()).toEqual({
      snapshot: 108 * mib,
      total: 156 * mib,
      archive: 172 * mib,
    });
    expect(captureArchiveLimitsFor({ snapshotSizeMultiplier: 3 })).toEqual({
      snapshot: 324 * mib,
      total: 372 * mib,
      archive: 388 * mib,
    });
    expect(captureArchiveLimitsFor({ snapshotSizeMultiplier: 0 })).toEqual(captureArchiveLimitsFor());
    expect(captureArchiveLimitsFor({ snapshotSizeMultiplier: 11 })).toEqual({
      snapshot: 1080 * mib,
      total: 1128 * mib,
      archive: 1144 * mib,
    });
  });

  it('round-trips Kotlin-compatible v1 and v2 payloads', async () => {
    const codec = new CaptureArchiveCodec();
    const v1 = await target('v1.apinspect');
    await codec.write(v1, metadata, { snapshotJson: '{"protocolVersion":{"major":1,"minor":1}}', screenshotPng: Buffer.from([1, 2]), rawArtifacts: { zip: Buffer.from([0x50, 0x4b]), text: 'raw' } });
    const readV1 = await codec.read(v1);
    expect(readV1.archiveVersion).toBe(1);
    expect(readV1.payload.rawArtifacts?.text).toBe('raw');

    const v2 = await target('v2.apinspect');
    const optionalPayloads = {
      analysisReportJson: '{"metrics":{"nodeCount":1,"maxDepth":1,"widestLevel":1},"findings":[]}',
      aiAnalysisReportJson: '{"model":"gpt-test","summary":"ok","findings":[]}',
      timelineHistoryJson: '{"frames":[{"index":0,"capturedAtEpochMillis":1}]}',
    };
    await codec.write(v2, metadata, { snapshotJson: '{}', composeInspectionJson: '{"nodes":[]}', ...optionalPayloads });
    const readV2 = await codec.read(v2);
    expect(readV2.archiveVersion).toBe(2);
    expect(readV2.payload.composeInspectionJson).toBe('{"nodes":[]}');
    expect(readV2.payload.analysisReportJson).toBe(optionalPayloads.analysisReportJson);
    expect(readV2.payload.aiAnalysisReportJson).toBe(optionalPayloads.aiAnalysisReportJson);
    expect(readV2.payload.timelineHistoryJson).toBe(optionalPayloads.timelineHistoryJson);
  });

  it('rejects unsafe paths, duplicate entries, and files absent from the manifest', async () => {
    const codec = new CaptureArchiveCodec();
    const unsafe = await target('unsafe.apinspect');
    const snapshot = Buffer.from('{}');
    await writeArchive(unsafe, [{ path: '../evil.json', bytes: snapshot, required: true }]);
    await expect(codec.read(unsafe)).rejects.toBeInstanceOf(CaptureArchiveFormatError);

    const duplicate = await target('duplicate.apinspect');
    const declared = [{ path: CAPTURE_ARCHIVE_PATHS.snapshot, bytes: snapshot, required: true }];
    await writeFile(duplicate, storedZip([
      { name: 'manifest.json', bytes: manifest(declared) },
      { name: CAPTURE_ARCHIVE_PATHS.snapshot, bytes: snapshot },
      { name: CAPTURE_ARCHIVE_PATHS.snapshot, bytes: snapshot },
    ]));
    await expect(codec.read(duplicate)).rejects.toThrow(/duplicate/);

    const undeclared = await target('undeclared.apinspect');
    await writeFile(undeclared, storedZip([
      { name: 'manifest.json', bytes: manifest(declared) },
      { name: CAPTURE_ARCHIVE_PATHS.snapshot, bytes: snapshot },
      { name: 'extra.txt', bytes: Buffer.from('extra') },
    ]));
    await expect(codec.read(undeclared)).rejects.toThrow(/manifest/);
  });

  it('rejects integrity mismatches, unsupported versions, and incomplete raw pairs', async () => {
    const codec = new CaptureArchiveCodec();
    const snapshot = Buffer.from('{}');
    const integrity = await target('integrity.apinspect');
    const declared = [{ path: CAPTURE_ARCHIVE_PATHS.snapshot, bytes: snapshot, required: true }];
    await writeFile(integrity, storedZip([
      { name: 'manifest.json', bytes: manifest(declared) },
      { name: CAPTURE_ARCHIVE_PATHS.snapshot, bytes: Buffer.from('[]') },
    ]));
    await expect(codec.read(integrity)).rejects.toThrow(/integrity/);

    const version = await target('version.apinspect');
    await writeArchive(version, declared, 3);
    await expect(codec.read(version)).rejects.toThrow(/version 3/);

    const raw = await target('raw.apinspect');
    await writeArchive(raw, [
      ...declared,
      { path: CAPTURE_ARCHIVE_PATHS.rawZip, bytes: Buffer.from('PK') },
    ]);
    await expect(codec.read(raw)).rejects.toThrow(/incomplete/);
  });

  it('rejects an entry declared above the configurable snapshot limit before integrity validation', async () => {
    const codec = new CaptureArchiveCodec();
    const path = await target('oversized.apinspect');
    const snapshot = Buffer.from('{}');
    const manifestBytes = Buffer.from(JSON.stringify({
      format: 'agentperf-inspector-capture', archiveVersion: 1, ...metadata,
      entries: [{
        path: CAPTURE_ARCHIVE_PATHS.snapshot,
        size: 108 * 1024 * 1024 + 1,
        sha256: createHash('sha256').update(snapshot).digest('hex'),
        required: true,
      }],
    }));
    await writeFile(path, storedZip([
      { name: CAPTURE_ARCHIVE_PATHS.manifest, bytes: manifestBytes },
      { name: CAPTURE_ARCHIVE_PATHS.snapshot, bytes: snapshot },
    ]));
    await expect(codec.read(path)).rejects.toThrow(/size is out of bounds/);
  });

  it('omits an oversized optional raw pair without writing an incomplete archive', async () => {
    const codec = new CaptureArchiveCodec();
    const path = await target('raw-too-large.apinspect');
    const result = await codec.write(path, metadata, {
      snapshotJson: '{}',
      rawArtifacts: { zip: Buffer.from('PK'), text: 'x'.repeat(8 * 1024 * 1024 + 1) },
    });
    expect(result.rawArtifactsIncluded).toBe(false);
    expect((await codec.read(path)).payload.rawArtifacts).toBeUndefined();
  });

  it('preserves an existing target when replacement fails and removes the temporary file', async () => {
    const path = await target('existing.apinspect');
    await writeFile(path, 'existing');
    const codec = new CaptureArchiveCodec({}, async () => { throw new Error('move failed'); });
    await expect(codec.write(path, metadata, { snapshotJson: '{}' })).rejects.toThrow('move failed');
    expect(await readFile(path, 'utf8')).toBe('existing');
  });
});
